Set `__doctest_requires__` in requires decorator

lagru opened the pull request:
## Description

Supersedes #7744.

This gets rid of the noise in docstrings introduced by the testsetup directive which users might not now about.

## Checklist

<!-- Before pull requests can be merged, they should provide: -->

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpydoc/blob/main/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

For maintainers and optionally contributors, please refer to the [instructions](https://scikit-image.org/docs/stable/development/contribute.html#documenting-changes) on how to document this PR for the release notes.

```release-note
...
```

lagru committed:
Set __doctest_requires__ in requires decorator

This gets rid of the noise in docstrings introduced by the testsetup
directive which users might not now about.

stefanv commented:
What do you think?

I think it's a bit odd to have version specifiers in there, when we already set minimums in the pyproject.toml, but that's a minor concern.

lagru commented:
> I think it's a bit odd to have version specifiers in there, when we already set minimums in the pyproject.toml, but that's a minor concern.

The version specifiers weren't added in this PR. I just made the parameter keyword-only. I don't feel strongly about it but as long as that option isn't a maintenance burden, I don't feel the need to remove it.

lagru committed:
Merge branch 'main' into require-doctest-skip

lagru committed:
Avoid unnecessary inline imports

I think this optimization isn't effective. As soon as any function is
decorated with this, functools is imported. Probably a relic from
before we added lazy loading.

lagru committed:
Improve doctest handling in require decorator

The assert could theoretically fail, if a function was wrapped twice
with this decorator. `inspect` is also the cleaner, more reliable
approach.

lagru committed:
Remove outdated __doctest_requires__ markers

these are now set by the decorator.

stefanv commented:
Could be useful; want to revisit?

lagru committed:
Merge branch 'main' into require-doctest-skip

Conflicts:
src/skimage/io/_io.py
src/skimage/restoration/_denoise.py

lagru committed:
Improve docstrings

lagru commented on src/_skimage2/_shared/version_requirements.py:
```diff
@@ -33,18 +35,20 @@ def get_module_version(module_name):
 
 
+@functools.cache
 def is_installed(name, version=None):
```
Should we update this function to use 
[`importlib.metadata.version`](https://docs.python.org/3/library/importlib.metadata.html#importlib.metadata.version). I'm guessing that's more robust than looking for a `__version__` attribute?

blink1073 replied:
I think you need to look for both, based on discussion I've seen on the Packaging category in discourse.

lagru replied:
Okay, I think currently, if `__version__` is not present, the version requirement is quietly ignored. I'd prefer it if it would emit a warning at the very least!

lagru replied:
That is
```python
is_installed("_skimage2._shared")  # has no version attribute
```
should return `False`.

blink1073 replied:
No, I mean that either one should suffice, with `__version__` being preferred if it is there.

lagru replied:
Okay, then https://github.com/scikit-image/scikit-image/pull/7814/commits/2c523b80b60f00c4223c103c283284460d2f6324 should be good enough to address my issue with a missing `__version__` being silently treated as success.

blink1073 replied:
So you don't want to add `importlib` handling here?  It seems we don't need it at the moment.

lagru replied:
Correct. With https://github.com/scikit-image/scikit-image/commit/2c523b80b60f00c4223c103c283284460d2f6324 making the failure (missing version attribute) more transparent, I'm comfortable leaving `importlib.metadata` for when we actually need it. :)

lagru commented on src/_skimage2/_shared/version_requirements.py:
```diff
@@ -75,39 +79,52 @@ def is_installed(name, version=None):
-    # functools
-    import functools
 
     def decorator(obj):
```
Should we simplify this bit with a [`contextmanager`](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager:~:text=contextmanager()%20uses%20ContextDecorator%20so%20the%20context%20managers%20it%20creates%20can%20be%20used%20as%20decorators%20as%20well%20as%20in%20with%20statements.)?

blink1073 replied:
I'd say the existing pattern is fine

lagru committed:
Skip blocking function call in doctest

`manual_polygon_segmentation` calls `plt.show(block=True)` internally.
We don't want this to be called and block even when maptlotlib is
available.

Also drop forgotten editing artifact.

lagru committed:
Fix outdated import for `_shared`

lagru committed:
Add missing `@xfail_without_pywt` markers

lagru committed:
Make `is_installed` more robust

That is, return ``False`` if no version attribute is available instead
of falling back to ``True``. This should help avoid confusing errors.

lagru committed:
Avoid warnings raised by skimage2 in doctests

blink1073 reviewed (approved):
LGTM
